libsolidity/codegen/ArrayUtils.cpp

```diff
@@ -774,6 +774,98 @@ void ArrayUtils::resizeDynamicArray(ArrayType const& _typeIn) const
 	);
 }
 
+void ArrayUtils::popStorageArrayElement(ArrayType const& _type) const
+{
+	solAssert(_type.location() == DataLocation::Storage, "");
+	solAssert(_type.isDynamicallySized(), "");
+	if (!_type.isByteArray() && _type.baseType()->storageBytes() < 32)
+		solAssert(_type.baseType()->isValueType(), "Invalid storage size for non-value type.");
+
+	if (_type.isByteArray())
+	{
+		// <ref> <byte_number> = <base_ref + index / itemsPerSlot> <(index % itemsPerSlot) * byteSize>
+		m_context.appendInlineAssembly(R"({
+			let slot_value := sload(ref)
+			switch and(slot_value, 1)
+			case 0 {
+				// short byte array
+				let length := and(div(slot_value, 2), 0x7f)
+				if iszero(length) { invalid() }
+				// Zero-out the suffix of the byte array by masking it
+				// Do not zero-out the least significant byte
+				// ((1<<(8 * (31 - length))) - 1) << 8
+				let mask := mul(0x100, sub(exp(0x100, sub(31, length)), 1))
+				slot_value := and(not(mask), slot_value)
+				// Reduce the length by 1
+				slot_value := sub(slot_value, 2)
+				sstore(ref, slot_value)
+			}
+			case 1 {
+				// long byte array
+				let length := div(slot_value, 2)
+
+				switch length
+				case 32
+				{
+					let slot := 0
+					let offset := 31
+
+					mstore(0, ref)
+					slot := add(keccak256(0, 0x20), slot)
+					let data := sload(slot)
+					sstore(slot, 0)
+					data := and(data, not(0xFF))
+					sstore(ref, or(data, 62))
+				}
+				default
+				{
+					let slot := div(sub(length, 1), 32)
+					let offset := and(sub(length, 1), 0x1F)
+
+					mstore(0, ref)
+					slot := add(keccak256(0, 0x20), slot)
+					let data := sload(slot)
+
+					// Zero-out the suffix of the byte array by masking it
+					// Do not zero-out the least significant byte
+					// ((1<<(8 * (31 - offset))) - 1)
+					let mask := sub(exp(0x100, sub(31, offset)), 1)
```

Comment:
I'm not sure anymore whether 31 is correct here. I think it should be 32.

Reply:
32 is correct :) This was also wrong for "short" byte arrays and led to the issue I was looking into earlier.